### DISCUSSION TITLE
[Proposal] Consider adding Fluent API / Markup style methods for the types defined in the Toolkit similar to that of existing Markup package for the Core

### DISCUSSION THREAD

**Discussion by egvijayanand** [CommunityToolkit/Maui#262] [New Feature Discussions]
# Fluent API / Markup style methods for the types defined in the Toolkit, wherever feasible

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
  * [ ] iOS Support
  * [ ] Android Support
  * [ ] macOS Support
  * [ ] Windows Support
* [ ] Unit Tests: Not Started
* [ ] Sample: Not Started
* [ ] Documentation: Not Started

## Summary
[summary]: #summary

Fluent API / Markup style methods for the types defined in the Toolkit, wherever feasible, are productivity boosters especially for those developing mainly with C# and coming from other backgrounds with no or less XAML knowledge.

## Motivation
[motivation]: #motivation

Already there exists a Markup package with support for the core features but there is none for the types in the Toolkit.

## Detailed Design
[design]: #detailed-design

We can start with adding fluent types for basic features such as Semantics, Attached Properties, and keep adding them incrementally to get up to speed.

## Usage Syntax
[usage]: #usage-syntax

In the below code sample, `SemanticHeading()` and `SemanticDesc()` is proposed fluent APIs. Similar ones need to be explored and added to the Toolkit.

Note: Since Semantics is now included as a core feature in .NET MAUI, these specific APIs need to be added to the `CommunityToolkit.Maui.Markup` package and shown as a sample to describe the proposal.

```CS
new Label()
{
	Text = "Welcome to .NET Multi-platform App UI",
}.Row(1).FontSize(18).CenterHorizontal().SemanticHeading(SemanticHeadingLevel.Level1).SemanticDesc("Welcome to dot net Multi platform App U I")
```

## Drawbacks
[drawbacks]: #drawbacks

The APIs need to be refreshed whenever there is a design change.

## Alternatives
[alternatives]: #alternatives

The classic approach of creating types with C# with instances and linking them. Have tried a similar thing for defining a Shell page in both the approaches (with and without such APIs). 

Adding a few such simple methods resulted in saving hundreds of keystrokes and code became much more readable.

## Unresolved Questions
[unresolved]: #unresolved-questions

The scope of the APIs that need to be added for the available types.

If any automation is possible, can be done to scan the types, generate those APIs, verify, and include them in the package.

**Comment by TheCodeTraveler** [comment]
Thanks @egvijayanand!

I’ve converted this to a Discussion because we can’t open a Proposal until we have a Detailed Design and API surface completed. 

This discussion may be better suited for CommunityToolkit.Maui.Markup: https://github.com/communitytoolkit/maui.markup

**Comment by egvijayanand** [comment]
Thanks, @brminnick.

The reason why I opened this issue here in this repo is since the Markup package is for the Core features and the request is for types in the CommunityToolkit.

Let me know your thoughts.

**Reply by pictos** [reply]
First of all, thanks for the proposal @egvijayanand❣
I agree with @brminnick, this discussion is from markup. We can add a reference to the CommunityToolkit.MAUI on CommunityToolkit.Markup and be able to create a fluent API for our features.

**Reply by egvijayanand** [reply]
@pictos I think CommunityToolkit.Maui.Markup package can continue as a standalone package as adding the Toolkit package as its dependency will force Toolkit even if developers don't use it.

The markup methods for the Toolkit can be added in this package itself as extension methods in a separate folder without any new dependency injected. This way existing users are can continue to use this package without any impact.

Others can add their inputs.

Regards,
Vijay Anand E G

**Reply by pictos** [reply]
@egvijayanand the `Maui.Markup` is the package where we host the Fluent API doesn't make sense to have those outside that package. It's fine to have a reference on `CommunityToolkit.MAUI` since `Maui.Markup` is a bunch of `static` APIs it can be easily trimmable, so devs will just pay for what they use.

**Reply by egvijayanand** [reply]
It's the other way around right, now `CommunityToolkit.Maui` will become a dependency for `CommunityToolkit.Maui.Markup` and the former package has both static and instance types, or am I missing something here.

**Reply by egvijayanand** [reply]
Moreover, on Xamarin both the packages are hosted in the same GitHub repo, hence there's a `Markup` package release for every `Toolkit` package release with identical release \# even though there's no change in the APIs for `Markup`.

Now I see they're into separate repos for Maui, the decision on this proposal also depends on how the release is planned in the future.

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
